### ISSUE TITLE
Remove access to undocumented check properties

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#2451]
As stated in PR #2450, there are several checks that have valid properties, not documented by us, that can be used in CS. Access to these properties shouldn't be allowed by the users as they were never meant to be used by them.

**Example:**
> ReturnCountCheck
http://checkstyle.sourceforge.net/config_coding.html#ReturnCount
> 
It actually has an undocument property, "compileFlags", coming from "AbstractFormatCheck".
The following config is valid, showing that this is indeed a valid property that CS will accept, but there is no valid reason the users should use it or see it documented.
````
	<module name="TreeWalker">
		<module name="ReturnCount">
			<property name="compileFlags" value="2" />
		</module>
	</module>
````

Once the PR is done, we should review the list of undocumented properties and come to a conclusion on how to remove user access to them.

**Possibilities:**
Remove the bean's setter method.
Make beans protected and access them only using "this.", also removing the setter method.
Set the beans' value via constructor.
**Last Resort:** remove extended class and only extend from Check, meaning duplicating many lines of code.

**Comment by romani** [comment]
no need to refactor setters usage, the less Hierarchy for Check the better, abstract functionally from "Check.java" class is enough.

**Comment by rnveach** [comment]
> abstract functionally from "Check.java" class is enough

This applies to all checks?
MissingCtorCheck and MissingSwitchDefaultCheck extends DescendantTokenCheck. Since DescendantTokenCheck isn't abstract, I can see why we should possible do it here.
AbstractFormatCheck is extended by 10 different checks. This is one of the main classes with undocumented properties.
I assume AbstractJavadocCheck is excluded? It is the heart of our javadoc checks.

So the end result will be no abstract classes under the "checks" folder, besides AbstractJavadocCheck, and all checks will extend Check or AbstractFileSetCheck?

Removing heirarchy would mean we have to duplicate code. If we ever found a bug in the duplicated part of the code, we would have to remember to apply it to all other places or run similar unit tests on them.

**Comment by romani** [comment]
Checkstyle is 15 years old project, the whole history of OOP is in it. On early 2000 every one tried to create hierarchy of classes to describe model and have minimum of code copy paste. Right now most of engineers understand that it was wrong philosophy, but code is written and works, so nobody refactor it.

But it do have problems:
- abstract classes introduce behaviour that Check does not need. And there is not way to hide it.
- usage of abstract classes increase dependency between Checks, you need to change Check and it looks like it should be in abstract class, and now you afraid to update Abstract class as you are not sure about new methods usage in all other Checks, so that will force you to keep new functionality in certain Check. And now a questions: What for you need such a Abstract ?
- In our case  original authors tries to minimized "copy-paste" that is minimal, and introduced a lot of dependencies between Checks, sometime unrelated Checks completely.

Ideally we should have only 3 Abstract classes: Check (on day I will rename it to AbstractCheck :) ), AbstractFileSetCheck and AbstractJavadocCheck. All these classes have no detail of validation process, just provide common events and proxing and .....

All Abstract classes except for these 3 should be removed, as they contain about 1 field for regexp. And some Check extends this Abstract and add one more field with regexp in Check implementation that create weird design that one RegExp filed is in abstract class, another is in certain Check ..... ehhhh.

It is completely ok to have minimal copy-paste in this world, whole life is structured on duplication with no relationship to each other. The less (reasonable less) dependencies the easier structure the less bugs in it, the more stable and reliable implementation. No need to join two classes if they occasionally (only for now) have similar fields.

Lets review each abstract class separately, to let them review one more time and be sure are on right way to remove it.

**Comment by romani** [comment]
@rnveach , To avoid wasting efforts for code changes for Abstracts lets discuss them before code changes.
I think that you have list of them at - https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java#L68

please propose a list of Abstracts that you would like to remove and keep.

**Comment by rnveach** [comment]
@romani 
We have already deprecated AbstractFormatCheck, AbstractIllegalCheck, AbstractIllegalMethodCheck, and AbstractDeclarationCollector. We have already decided to keep using AbstractNameCheck.
The following are the ones we haven't touched yet.

I feel the following should be deprecated. They don't provide us much benefit.
* AbstractAccessControlNameCheck, AbstractClassCouplingCheck, AbstractComplexityCheck, AbstractNestedDepthCheck, AbstractOptionCheck, AbstractTypeParameterNameCheck.

The following have quite a few methods, so I am unsure how we can deprecate them. Maybe they could be made into utility classes or something else?
* AbstractHeaderCheck, AbstractParenPadCheck, AbstractSuperCheck

AbstractTypeAwareCheck is already deprecated but it is still used by JavadocMethodCheck which suppresses the deprecation. I am unsure what to do with this one since we don't support type aware checks.

**Comment by romani** [comment]
AbstractAccessControlNameCheck extends AbstractNameCheck, lets keep it for now, as we keep AbstractNameCheck.

ClassDataAbstractionCouplingCheck and ClassFanOutComplexityCheck extends AbstractClassCouplingCheck, looks like they have whole logic as common, deprecation will result serious code duplication. Lets keep AbstractClassCouplingCheck for now.

AbstractComplexityCheck - we need to deprecate.

AbstractNestedDepthCheck - we need to deprecate.

AbstractOptionCheck - we need to deprecate.

AbstractTypeParameterNameCheck - we need to deprecate, completely useless.

AbstractHeaderCheck, AbstractParenPadCheck, AbstractSuperCheck - lets keep them for now.

> AbstractTypeAwareCheck is already deprecated but it is still used by JavadocMethodCheck which suppresses the deprecation. I am unsure what to do with this one since we don't support type aware checks.

All JavadocChecks that are not based on  AbstractJavadocCheck.java should be reimplemented to use AbstractJavadocCheck.java as base class and do not use parsing by RegExp. All Javadoc checks should use ANTLR grammar. https://github.com/checkstyle/checkstyle/wiki/Checkstyle-GSoC-2014-Project-Ideas#project-name-antlr-grammar-for-comments-and-documentation-comments-javadocs .

**Comment by rnveach** [comment]
> AbstractAccessControlNameCheck ..., lets keep it for now

Are we keeping the undocumented properties or will we document them? Or make the check in question extend AbstractNameCheck?
https://github.com/checkstyle/checkstyle/blob/35ecc0b03fc1b4ad5454873c77a8b2694ee92e74/src/test/java/com/puppycrawl/tools/checkstyle/XDocsPagesTest.java#L114-L117

**Comment by romani** [comment]
> Are we keeping the undocumented properties or will we document them? Or make the check in question extend AbstractNameCheck?

http://checkstyle.sourceforge.net/config_naming.html#MethodName
Please document these option, they are pretty reasonable - https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbstractAccessControlNameCheck.java#L50.

**Comment by rnveach** [comment]
@romani 
For the most part this issue is done. 

I haven't touched the `JavadocMethodCheck` extending `AbstractTypeAwareCheck` since this probably requires a rewrite or at least a more in-depth analysis to see how it can be moved to extend `AbstractJavadocCheck`.
If it is ok, we can move this to its own issue that maybe someone else can work on.

As for the others, I looked over all the extends and didn't see any left that you said we should remove right now.

**Comment by romani** [comment]
thanks a lot!

> JavadocMethodCheck

It will be reimplemented, no need to do this now.
